A40-csds-support.md

```diff
@@ -0,0 +1,275 @@
+
+Here is an example of NACK information handling with new ADS changes:
+
+```
+# Imagine we have endpoint A, B, C
+EDS -> {A, B, C}, version 1
+gRPC -> ACK
+CSDS -> [
+  {Endpoint A, ACK, version 1},
+  {Endpoint B, ACK, version 1},
+  {Endpoint C, ACK, version 1}
+]
+
+# The newer endpoint B contains parsing error
+EDS -> {A, B}, version 2
+gRPC -> NACK, Failed to parse endpoint B 
+CSDS -> [
+  {Endpoint A, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint B, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint C, ACK, version 1}
+]
+
+# Accepted update will clean error states
+EDS -> {B, C} version 3
+gRPC -> ACK
+CSDS -> [
+  {Endpoint A, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint B, ACK, version 3},
+  {Endpoint C, ACK, version 3}
+]
+```
+
+
+### xDS Config Error State
+
+When an ADS response is rejected, gRPC should provide debug information via
+CSDS. This is done via `UpdateFailureState` within the `config_dump.proto`,
+which includes the version, timestamp, and the reason of the rejected update
+(see [proto
+definition](https://github.com/envoyproxy/envoy/blob/a4024a578b3f2611fe26229f5d0de99eb0c56895/api/envoy/admin/v3/config_dump.proto#L72)):
+
+
+### CSDS Service Design
+
+Ideally, the CSDS service should be reusable by our users, instead of just an
+internal tool for the gRPC library. So, the CSDS service should be implemented
+in wrapper languages and Java/Go. There are three major parts in the service
+design:
+
+* Cache resources in the global XdsClient;
+* Collect additional metadata about the cached resource (update timestamp,
+  version, etc.);
+* Assemble cached resources into the gigantic config dump response.
+
+Note that xDS config should include the `config.core.v3.Node` information in the
+`envoy.service.status.v3.ClientConfig.node` field. Unlike other xDS configs,
+this information is static and may require extra logic to integrate into the
+CSDS service.
+
+For Java and Go, it’s straightforward that this feature can be implemented
+bottom-up in the same language, and it’s possible to avoid an additional copy of
+the message itself. But Core needs to transport the collected xDS configs across
+the language boundary. The API needs to convert proto messages into C-compatible
+types (e.g. `char *`, either bytes or JSON).
+
+
+#### Detail: No xDS v2 Support
+
+CSDS was meant to be a RPC service made for xDS management servers before this
+proposal. But the service itself has the potential to serve client status on xDS
+clients. However, during development, we found several constraints about the
+existing service protocol, hence we merged several updates (see above) to
+improve the CSDS service to meet our standard. The updates are made to xDS v3
+and xDS v4 (alpha) only, since xDS v2 is in deprecated state. This doc proposes
+to not support xDS v2 for CSDS.
```

Comment:
What to do if the underlying client is speaking v2?
Return an error? What error code to use? `FailedPrecondition`?

Reply:
By underlying client, do you mean the gPRC application? I thought gRPC processes all xDS config into v3, so the dumped xDS config is agnostic to v2/v3. If the external CSDS client is only speaking v2, they won't be able to invoke the v3 method, because the fully-qualified-method-name includes the service name which has "v3" in it. So, the v2 external CSDS client will get UNIMPLEMENTED by default, in theory.

Reply:
Here's what TD does for v3 (currently only v2 supported):

```
❯ csds-client -request_file request.yaml -api_version v3
2021/03/01 18:20:55 rpc error: code = Unimplemented desc = The GRPC target is not implemented on the server, host: trafficdirector.googleapis.com, method: /envoy.service.status.v3.ClientStatusDiscoveryService/StreamClientStatus.
```

Reply:
Nice, so in practicce, the old client will also get UNIMPLEMENTED.

Reply:
Note that the CSDS RPC service name is different between v2 and v3, because the major version is part of the proto package name.  So I think all we need to do here is to implement only the v3 service, not the v2 service.  Then clients will get UNIMPLEMENTED if they try to use the v2 service, just like they would for any other unimplemented RPC service.

Reply:
Oh, sorry, Lidi already said exactly what I just said. :)

Reply:
So the answer is, we support csds v3 only, csds v2 will get error with code unimplemented.
But the data can be either xds v2 or xds v3, right?

Reply:
If the underlying data is xds v2, that means the `proto.Any`s in the xds responses are v2 resources.

And in csds, I would assume the dumped resources (`proto.Any`s) should be v3 type? That means we need to turn those v2 `proto.Any` resources from the xds responses into v3 `proto.Any` (unmarshal and then marshal again), and then put in the dumped messages, is that right?

(We alreay unmarshal the v2 resources into v3 types to read. So this basically means we need to then marshal the v3 type into `proto.Any` and keep those in cache as raw resources, instead of the `proto.Any`s from the xds responses. It's not too many extra steps. But just want to confirm this.)

Reply:
I would vote for yes, let's convert the packed v2 resources into v3.

It would make analyzing the CSDS response simpler. Though, I can only think of `proto.Any` for ApiListener and HTTP filters. ApiListener will be converted to v3 for sure. We don't support v2 xDS HTTP filters at all, so they will be rejected in the first place. Do you have an example that where the v2 resources might slip through?

Reply:
I think it's fine to send v2 protos inside of the `Any` fields in v3 CSDS.  In xDS terms, the transport protocol version and the data model version do not have to be the same.  I don't think we need to do any conversion here.

Reply:
> Do you have an example that where the v2 resources might slip through

It's not one type of resource. All the resources in xds v2 are v2. E.g. `v2.Cluster` vs `v3.Cluster`, `v2.Endpoint` vs `v3.Endpoint`.

> In xDS terms, the transport protocol version and the data model version do not have to be the same.

That's right.
The difference would be on the csds client side. If we send v2 protos as `Any` in the csds responses, the csds client needs to depend on v2 protos (not that they will be used, but their types need to be registered) so the client can unmarshal those `Any`.

Reply:
I think it's fine for CSDS client to need v2 protos if the client is using v2.  That's independent of the fact that we're only supporting v3 of the CSDS RPC service itself.

This is no different than ADS -- it's possible for a client to use v3 of ADS but v2 of the resource protos, or vice versa.  In fact, in Envoy, those are two independent knobs.  In gRPC, we just simplified it and avoided one of the knobs by depending on the fact that we didn't happen to use any fields in v2 that were being removed in v3, so we could just unconditionally treat all of the resource protos as v3.

Reply:
This SGTM (and is actually easier to do :))

Let's update the gRFC to clarify this? This doesn't sound accurate:
> `This doc proposes
to not support xDS v2 for CSDS.`

Reply:
Updated to
> For xDS CSDS v2 (the RPC service itself), this doc proposes to not support, so
> we only provide v3 CSDS service and CSDS client requesting for v2 service will
> get UNIMPLEMENTED error.
> 
> There are edge cases that v2 and v3 xDS resources are mixed together. This doc
> proposes to keep them as-is in the CSDS responses. So, the CSDS service can
> accurately reflect the xDS configs received from the control plane.

The section title also updated to `xDS v2/v3/v4`, since it's hard to describe our handling of v2 in the title.